### PR TITLE
Reflect changes to server selector in opaqueness

### LINKED ISSUE #11995: Changes to server selector are not reflected in opaqueness
### What is the issue?

If a Server is marking a Pod's port as opaque and then the Server's podSelector is updated to no longer select that Pod, then the Pod's port should no longer be marked as opaque.  However, this update does not result in any updates from the destination API's `Get` stream and the port remains marked as opaque.

### How can it be reproduced?

1. Create a Pod, a Service which selects that pod, and a Server which selects the pod and marks a port as opaque
2. Initiate a `Get` stream for the service and notice that the endpoint is marked with opaque transport:

```
go run controller/script/destination-client/main.go -method get -path test.emojivoto.svc.cluster.local:8080  --token '{"nodeName":"alex-worker"}'
INFO[0000] Add:
INFO[0000] labels: map[namespace:emojivoto service:test]
INFO[0000] - 10.42.0.163:8080
INFO[0000]   - labels: map[control_plane_ns:linkerd deployment:test pod:test-5d75589c45-wv5kz pod_template_hash:5d75589c45 serviceaccount:default zone:]
INFO[0000]   - protocol hint: UNKNOWN
INFO[0000]   - identity: dns_like_identity:{name:"default.emojivoto.serviceaccount.identity.linkerd.cluster.local"}  server_name:{name:"default.emojivoto.serviceaccount.identity.linkerd.cluster.local"}
INFO[0000]   - opaque transport port: 4143
```

3. Edit the Server's selector so that it no longer selects the Pod
4. Notice no updates on the `Get` stream.

### Logs, error output, etc

This is because we ignore Server updates if the Server's selector does not select the pod:

https://github.com/linkerd/linkerd2/blob/main/controller/api/destination/watcher/endpoints_watcher.go#L1251

### output of `linkerd check -o short`

N/A

### Environment

reproduced on commit: 796bb8532336796e153968e643aba1a06f62773c
k3d version v5.6.0
k3s version v1.27.4-k3s1 (default)

### Possible solution

_No response_

### Additional context

_No response_

### Would you like to work on fixing this bug?

None

### PR DESCRIPTION
**Pull request by adleong**
Fixes #11995

If a Server is marking a Pod's port as opaque and then the Server's podSelector is updated to no longer select that Pod, then the Pod's port should no longer be marked as opaque. However, this update does not result in any updates from the destination API's Get stream and the port remains marked as opaque.

We fix this by updating the endpoint watcher's handling of Server updates to consider both the old and the new Server.

### CODE REVIEW

File: controller/api/destination/server_test.go [hunk 1]
```diff
@@ -158,6 +159,110 @@ func TestGet(t *testing.T) {
...
+				Namespace: "ns",
+			},
+			Spec: v1beta2.ServerSpec{
+				PodSelector: &metav1.LabelSelector{
+					// PodSelector is updated to NOT select the pod
+					MatchLabels: map[string]string{"app": "FOOBAR"},
+				},
+				ExternalWorkloadSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{"app": "external-workload-policy-test"},
+				},
```
Comment (zaharidichev) [inline controller/api/destination/server_test.go:216]: what is the significance of having the ExternalWorkloadSelector selector set here in this test. Isn't there a one-of constraint present on the podSelector and externalWorkloadSelector fields in general

File: controller/api/destination/server_test.go [hunk 2]
```diff
@@ -158,6 +159,110 @@ func TestGet(t *testing.T) {
 		}
 	})
 
+	t.Run("Return endpoint opaque protocol controlled by a server", func(t *testing.T) {
```
Comment (zaharidichev) [inline controller/api/destination/server_test.go:162]: can we add the same test for external workloads ?

Comment (adleong) [reply controller/api/destination/server_test.go:163]: great question.  I tried this and then spent forever racking my brain trying to figure out why it wasn't working.  it turns out the server_tests all run with the endpoint_watcher in Endpoints mode, not EndpointSlices mode, and external workloads are not compatible with Endpoints.  

Given that EndpointSlices mode is the default, we'll want to rewrite these tests to use it, but not in this PR.

**Review by zaharidichev — APPROVED** [review]

File: controller/api/destination/watcher/endpoints_watcher.go [hunk 3]
```diff
@@ -1237,70 +1242,12 @@ func (pp *portPublisher) unsubscribe(listener EndpointUpdateListener) {
...
-			}
-
-		} else {
-			continue
-		}
-
-		if portMatch {
-			if isAdd && server.Spec.ProxyProtocol == opaqueProtocol {
+		if pp.isAddressSelected(address, oldServer) || pp.isAddressSelected(address, newServer) {
+			if newServer != nil && pp.isAddressSelected(address, newServer) && newServer.Spec.ProxyProtocol == opaqueProtocol {
```
Comment (mateiidavid) [inline controller/api/destination/watcher/endpoints_watcher.go:1250]: Could we maybe simplify this by saying:

```Go
if pp.isAddressSelected(address, oldServer) || pp.isAddressSelected(address, newServer) {
   if pp.isAddressSelected(address, newServer) && newServer.Spec.ProxyProtocol == opaqueProtocol {
   }
}
```

iiuc `isAddressSelected` will return false if the server it is applied to is nil so it seems redundant to check it in this if condition too.

**Review by mateiidavid — APPROVED** [review]
Looks good!

**Pull request merged**